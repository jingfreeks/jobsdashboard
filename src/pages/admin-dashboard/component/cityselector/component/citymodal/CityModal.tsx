import { memo } from "react";
import { X, Upload } from "lucide-react";

interface CityModalProps {
  isOpen: boolean;
  title: string;
  cityName: string;
  selectedStateId: string;
  selectedImage?: File | null;
  imagePreview?: string | null;
  states: Array<{ _id: string; name: string }> | undefined;
  onCityNameChange: (name: string) => void;
  onStateChange: (stateId: string) => void;
  onImageChange?: (file: File | null) => void;
  onImagePreviewChange?: (preview: string | null) => void;
  onSubmit: (e: React.FormEvent) => void;
  onClose: () => void;
  isLoading: boolean;
  isUploading?: boolean;
  submitText: string;
  showImageUpload?: boolean;
}

const CityModal = memo<CityModalProps>(({ 
  isOpen, 
  title, 
  cityName, 
  selectedStateId, 
  imagePreview, 
  states, 
  onCityNameChange, 
  onStateChange, 
  onImageChange, 
  onImagePreviewChange, 
  onSubmit, 
  onClose, 
  isLoading, 
  isUploading = false,
  submitText,
  showImageUpload = false
}) => {
  if (!isOpen) return null;

  return (
    <div className="fixed inset-0 z-50 flex items-center justify-center bg-black bg-opacity-40">
      <div className="bg-white rounded-xl shadow-lg p-8 w-full max-w-md relative">
        <button
          onClick={onClose}
          className="absolute top-3 right-3 text-slate-400 hover:text-slate-700 text-xl font-bold"
        >
          &times;
        </button>
        <h3 className="text-xl font-bold mb-4 text-slate-800">
          {title}
        </h3>
        <form onSubmit={onSubmit} className="flex flex-col gap-4">
          <input
            type="text"
            className="border border-slate-200 rounded-lg px-4 py-2 focus:outline-none focus:border-blue-500"
            placeholder="City Name"
            value={cityName}
            onChange={(e) => onCityNameChange(e.target.value)}
            autoFocus
            required
            disabled={isLoading}
          />
          <select
            className="border border-slate-200 rounded-lg px-4 py-2 focus:outline-none focus:border-blue-500"
            value={selectedStateId}
            onChange={(e) => onStateChange(e.target.value)}
            disabled={isLoading}
          >
            <option value="">Select State</option>
            {states?.map((state) => (
              <option key={state._id} value={state._id}>
                {state.name}
              </option>
            ))}
          </select>
          
          {/* Image Upload - only show for add modal */}
          {showImageUpload && onImageChange && onImagePreviewChange && (
            <div className="space-y-2">
              <label className="block text-sm font-medium text-slate-700">City Image</label>
              {imagePreview ? (
                <div className="relative">
                  <img
                    src={imagePreview}
                    alt="Preview"
                    className="w-full h-32 object-cover rounded-lg border border-slate-200"
                  />
                  <button
                    type="button"
                    onClick={() => {
                      onImageChange(null);
                      onImagePreviewChange(null);
                    }}
                    className="absolute top-2 right-2 bg-red-500 text-white rounded-full p-1 hover:bg-red-600"
                  >
                    <X className="w-4 h-4" />
                  </button>
                </div>
              ) : (
                <div className="border-2 border-dashed border-slate-300 rounded-lg p-6 text-center hover:border-blue-400 transition-colors">
                  <Upload className="mx-auto h-8 w-8 text-slate-400 mb-2" />
                  <input
                    type="file"
                    accept="image/*"
                    onChange={(e) => {
                      const file = e.target.files?.[0];
                      if (file) {
                        onImageChange(file);
                        const reader = new FileReader();
                        reader.onload = (e) => {
                          onImagePreviewChange(e.target?.result as string);
                        };
                        reader.readAsDataURL(file);
                      }
                    }}
                    className="hidden"
                    id="image-upload"
                    disabled={isLoading}
                  />
                  <label
                    htmlFor="image-upload"
                    className="cursor-pointer text-sm text-slate-600 hover:text-blue-600"
                  >
                    Click to upload image
                  </label>
                </div>
              )}
            </div>
          )}
          
          <div className="flex gap-2 justify-end">
            <button
              type="button"
              onClick={onClose}
              className="px-4 py-2 rounded-lg bg-slate-100 hover:bg-slate-200 text-slate-700 font-semibold"
              disabled={isLoading}
            >
              Cancel
            </button>
            <button
              type="submit"
              className="px-4 py-2 rounded-lg bg-blue-600 hover:bg-blue-700 text-white font-semibold disabled:opacity-50"
              disabled={isLoading || isUploading}
            >
              {isLoading || isUploading ? 'Processing...' : submitText}
            </button>
          </div>
        </form>
      </div>
    </div>
  );
});

CityModal.displayName = 'CityModal';

export default CityModal; 