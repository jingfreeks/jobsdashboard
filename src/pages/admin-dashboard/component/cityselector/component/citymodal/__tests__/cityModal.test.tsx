
import { render, screen, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi, beforeEach } from "vitest";
import CityModal from "../CityModal";

// // Mock FileReader for image preview testing
// class MockFileReader {
//   // eslint-disable-next-line @typescript-eslint/no-explicit-any
//   onload: ((this: FileReader, ev: ProgressEvent<FileReader>) => any) | null =
//     null;
//   readAsDataURL(file: Blob) {
//     if (this.onload) {
//       this.onload({
//         target: { result: "data:image/png;base64,mockdata" },
//       } as ProgressEvent<FileReader>);
//     }
//   }
// }
// (global as any).FileReader = MockFileReader;

describe("CityModal Component", () => {
  let defaultProps: any;

  beforeEach(() => {
    defaultProps = {
      isOpen: true,
      title: "Add City",
      cityName: "Sample City",
      selectedStateId: "1",
      states: [
        { _id: "1", name: "State One" },
        { _id: "2", name: "State Two" },
      ],
      onCityNameChange: vi.fn(),
      onStateChange: vi.fn(),
      onImageChange: vi.fn(),
      onImagePreviewChange: vi.fn(),
      onSubmit: vi.fn((e) => e.preventDefault()),
      onClose: vi.fn(),
      isLoading: false,
      isUploading: false,
      submitText: "Save",
    };
  });

  it("matches snapshot (with image upload)", () => {
    const { asFragment } = render(
      <CityModal {...defaultProps} showImageUpload={true} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot (without image upload)", () => {
    const { asFragment } = render(
      <CityModal {...defaultProps} showImageUpload={false} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("does not render when isOpen is false", () => {
    const { container } = render(
      <CityModal {...defaultProps} isOpen={false} />
    );
    expect(container.firstChild).toBeNull();
  });

  it("calls onClose when cancel button is clicked", () => {
    render(<CityModal {...defaultProps} />);
    fireEvent.click(screen.getByText(/Cancel/i));
    expect(defaultProps.onClose).toHaveBeenCalled();
  });

  it("calls onCityNameChange when typing", () => {
    render(<CityModal {...defaultProps} />);
    fireEvent.change(screen.getByPlaceholderText(/City Name/i), {
      target: { value: "New City" },
    });
    expect(defaultProps.onCityNameChange).toHaveBeenCalledWith("New City");
  });

  it("calls onStateChange when selecting state", () => {
    render(<CityModal {...defaultProps} />);
    fireEvent.change(screen.getByDisplayValue("State One"), {
      target: { value: "2" },
    });
    expect(defaultProps.onStateChange).toHaveBeenCalledWith("2");
  });

  it("handles view image file after upload", () => {
    const { asFragment } = render(
      <CityModal {...defaultProps} imagePreview={true} showImageUpload={true} onImageChange={true} onImagePreviewChange={true} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
  //   it("handles image file upload", () => {
  //     render(<CityModal {...defaultProps} imagePreview={true} showImageUpload={true} />);
  //     const file = new File(["dummy"], "city.png", { type: "image/png" });
  //     const input = screen.getByLabelText(/Click to upload image/i).previousSibling as HTMLInputElement;
  //     fireEvent.change(input, { target: { files: [file] } });

  //     expect(defaultProps.onImageChange).toHaveBeenCalledWith(file);
  //     expect(defaultProps.onImagePreviewChange).toHaveBeenCalledWith("data:image/png;base64,mockdata");
  //   });

  //   it("removes image when delete button clicked", () => {
  //     render(
  //       <CityModal
  //         {...defaultProps}
  //         showImageUpload={true}
  //         imagePreview="test.jpg"
  //       />
  //     );
  //     fireEvent.click(screen.getByRole("button", { name: /X/i }));
  //     expect(defaultProps.onImageChange).toHaveBeenCalledWith(null);
  //     expect(defaultProps.onImagePreviewChange).toHaveBeenCalledWith(null);
  //   });

  it("disables submit button when loading", () => {
    render(<CityModal {...defaultProps} isLoading={true} />);
    expect(screen.getByRole("button", { name: /Processing/i })).toBeDisabled();
  });

  it("submits form when save clicked", () => {
    render(<CityModal {...defaultProps} />);
    fireEvent.click(screen.getByText(/Save/i));
    expect(defaultProps.onSubmit).toHaveBeenCalled();
  });
});
