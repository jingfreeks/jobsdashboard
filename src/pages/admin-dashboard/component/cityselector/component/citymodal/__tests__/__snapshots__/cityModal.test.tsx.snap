// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CityModal Component > handles view image file after upload 1`] = `
<DocumentFragment>
  <div
    class="fixed inset-0 z-50 flex items-center justify-center bg-black bg-opacity-40"
  >
    <div
      class="bg-white rounded-xl shadow-lg p-8 w-full max-w-md relative"
    >
      <button
        class="absolute top-3 right-3 text-slate-400 hover:text-slate-700 text-xl font-bold"
      >
        ×
      </button>
      <h3
        class="text-xl font-bold mb-4 text-slate-800"
      >
        Add City
      </h3>
      <form
        class="flex flex-col gap-4"
      >
        <input
          class="border border-slate-200 rounded-lg px-4 py-2 focus:outline-none focus:border-blue-500"
          placeholder="City Name"
          required=""
          type="text"
          value="Sample City"
        />
        <select
          class="border border-slate-200 rounded-lg px-4 py-2 focus:outline-none focus:border-blue-500"
        >
          <option
            value=""
          >
            Select State
          </option>
          <option
            value="1"
          >
            State One
          </option>
          <option
            value="2"
          >
            State Two
          </option>
        </select>
        <div
          class="space-y-2"
        >
          <label
            class="block text-sm font-medium text-slate-700"
          >
            City Image
          </label>
          <div
            class="relative"
          >
            <img
              alt="Preview"
              class="w-full h-32 object-cover rounded-lg border border-slate-200"
              src="true"
            />
            <button
              class="absolute top-2 right-2 bg-red-500 text-white rounded-full p-1 hover:bg-red-600"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-x w-4 h-4"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M18 6 6 18"
                />
                <path
                  d="m6 6 12 12"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="flex gap-2 justify-end"
        >
          <button
            class="px-4 py-2 rounded-lg bg-slate-100 hover:bg-slate-200 text-slate-700 font-semibold"
            type="button"
          >
            Cancel
          </button>
          <button
            class="px-4 py-2 rounded-lg bg-blue-600 hover:bg-blue-700 text-white font-semibold disabled:opacity-50"
            type="submit"
          >
            Save
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`CityModal Component > matches snapshot (with image upload) 1`] = `
<DocumentFragment>
  <div
    class="fixed inset-0 z-50 flex items-center justify-center bg-black bg-opacity-40"
  >
    <div
      class="bg-white rounded-xl shadow-lg p-8 w-full max-w-md relative"
    >
      <button
        class="absolute top-3 right-3 text-slate-400 hover:text-slate-700 text-xl font-bold"
      >
        ×
      </button>
      <h3
        class="text-xl font-bold mb-4 text-slate-800"
      >
        Add City
      </h3>
      <form
        class="flex flex-col gap-4"
      >
        <input
          class="border border-slate-200 rounded-lg px-4 py-2 focus:outline-none focus:border-blue-500"
          placeholder="City Name"
          required=""
          type="text"
          value="Sample City"
        />
        <select
          class="border border-slate-200 rounded-lg px-4 py-2 focus:outline-none focus:border-blue-500"
        >
          <option
            value=""
          >
            Select State
          </option>
          <option
            value="1"
          >
            State One
          </option>
          <option
            value="2"
          >
            State Two
          </option>
        </select>
        <div
          class="space-y-2"
        >
          <label
            class="block text-sm font-medium text-slate-700"
          >
            City Image
          </label>
          <div
            class="border-2 border-dashed border-slate-300 rounded-lg p-6 text-center hover:border-blue-400 transition-colors"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-upload mx-auto h-8 w-8 text-slate-400 mb-2"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 3v12"
              />
              <path
                d="m17 8-5-5-5 5"
              />
              <path
                d="M21 15v4a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2v-4"
              />
            </svg>
            <input
              accept="image/*"
              class="hidden"
              id="image-upload"
              type="file"
            />
            <label
              class="cursor-pointer text-sm text-slate-600 hover:text-blue-600"
              for="image-upload"
            >
              Click to upload image
            </label>
          </div>
        </div>
        <div
          class="flex gap-2 justify-end"
        >
          <button
            class="px-4 py-2 rounded-lg bg-slate-100 hover:bg-slate-200 text-slate-700 font-semibold"
            type="button"
          >
            Cancel
          </button>
          <button
            class="px-4 py-2 rounded-lg bg-blue-600 hover:bg-blue-700 text-white font-semibold disabled:opacity-50"
            type="submit"
          >
            Save
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`CityModal Component > matches snapshot (without image upload) 1`] = `
<DocumentFragment>
  <div
    class="fixed inset-0 z-50 flex items-center justify-center bg-black bg-opacity-40"
  >
    <div
      class="bg-white rounded-xl shadow-lg p-8 w-full max-w-md relative"
    >
      <button
        class="absolute top-3 right-3 text-slate-400 hover:text-slate-700 text-xl font-bold"
      >
        ×
      </button>
      <h3
        class="text-xl font-bold mb-4 text-slate-800"
      >
        Add City
      </h3>
      <form
        class="flex flex-col gap-4"
      >
        <input
          class="border border-slate-200 rounded-lg px-4 py-2 focus:outline-none focus:border-blue-500"
          placeholder="City Name"
          required=""
          type="text"
          value="Sample City"
        />
        <select
          class="border border-slate-200 rounded-lg px-4 py-2 focus:outline-none focus:border-blue-500"
        >
          <option
            value=""
          >
            Select State
          </option>
          <option
            value="1"
          >
            State One
          </option>
          <option
            value="2"
          >
            State Two
          </option>
        </select>
        <div
          class="flex gap-2 justify-end"
        >
          <button
            class="px-4 py-2 rounded-lg bg-slate-100 hover:bg-slate-200 text-slate-700 font-semibold"
            type="button"
          >
            Cancel
          </button>
          <button
            class="px-4 py-2 rounded-lg bg-blue-600 hover:bg-blue-700 text-white font-semibold disabled:opacity-50"
            type="submit"
          >
            Save
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
